File: android/src/main/java/com/github/quadflask/react/navermap/RNNaverMapMarkerManager.java
package com.github.quadflask.react.navermap;

import android.content.Context;
import android.graphics.Color;
import android.os.Build;
import android.util.DisplayMetrics;
import android.view.WindowManager;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.annotations.ReactProp;

import static com.github.quadflask.react.navermap.ReactUtil.toNaverLatLng;

public class RNNaverMapMarkerManager extends EventEmittableViewGroupManager<RNNaverMapMarker> {
    private final DisplayMetrics metrics;

    public RNNaverMapMarkerManager(ReactApplicationContext reactContext) {
        super(reactContext);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            metrics = new DisplayMetrics();
            ((WindowManager) reactContext.getSystemService(Context.WINDOW_SERVICE))
                    .getDefaultDisplay()
                    .getRealMetrics(metrics);
        } else {
            metrics = reactContext.getResources().getDisplayMetrics();
        }
    }

    @Override
    String[] getEventNames() {
        return new String[]{
                "onClick"
        };
    }

    @NonNull
    @Override
    public String getName() {
        return "RNNaverMapMarker";
    }

    @NonNull
    @Override
    protected RNNaverMapMarker createViewInstance(@NonNull ThemedReactContext reactContext) {
        return new RNNaverMapMarker(this, reactContext);
    }

    @ReactProp(name = "coordinate")
    public void setCoordinate(RNNaverMapMarker view, ReadableMap map) {
        view.setCoordinate(toNaverLatLng(map));
    }

    @ReactProp(name = "anchor")
    public void setAnchor(RNNaverMapMarker view, ReadableMap map) {
        // should default to (0.5, 1) (bottom middle)
        float x = map != null && map.hasKey("x") ? (float) map.getDouble("x") : 0.5f;
        float y = map != null && map.hasKey("y") ? (float) map.getDouble("y") : 1.0f;
        view.setAnchor(x, y);
    }

    @ReactProp(name = "image")
    public void setImage(RNNaverMapMarker view, @Nullable String source) {
        view.setImage(source);
    }

    @ReactProp(name = "pinColor", defaultInt = Color.RED, customType = "Color")
    public void setPinColor(RNNaverMapMarker view, int pinColor) {
        view.setIconTintColor(pinColor);
    }

    @ReactProp(name = "rotation", defaultFloat = 0.0f)
    public void setMarkerRotation(RNNaverMapMarker view, float rotation) {
        view.setRotation(rotation);
    }

    @ReactProp(name = "flat", defaultBoolean = false)
    public void setFlat(RNNaverMapMarker view, boolean flat) {
        view.setFlat(flat);
    }

    @ReactProp(name = "width", defaultFloat = 64)
    public void setWidth(RNNaverMapMarker view, float width) {
        int widthInScreenPx = Math.round(metrics.density * width);
        view.setWidth(widthInScreenPx);
    }

    @ReactProp(name = "height", defaultFloat = 64)
    public void setHeight(RNNaverMapMarker view, float height) {
        int heightInScreenPx = Math.round(metrics.density * height);
        view.setHeight(heightInScreenPx);
    }

    @ReactProp(name = "animated", defaultBoolean = false)
    public void setAnimated(RNNaverMapMarker view, boolean animated) {
        view.setAnimated(animated);
    }

    @ReactProp(name = "easing", defaultInt = -1)
    public void setEasing(RNNaverMapMarker view, int easingFunction) {
        view.setEasing(easingFunction);
    }

    @ReactProp(name = "duration", defaultInt = 500)
    public void setDuration(RNNaverMapMarker view, int duration) {
        view.setDuration(duration);
    }

    @ReactProp(name = "alpha", defaultFloat = 1f)
    public void setAlpha(RNNaverMapMarker view, float alpha) {
        view.setAlpha(alpha);
    }
}

File: android/src/main/java/com/github/quadflask/react/navermap/RNNaverMapMarkerManager.kt
package com.github.quadflask.react.navermap

import android.graphics.Color
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.View
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.annotations.ReactProp
import com.github.quadflask.react.navermap.ReactUtil.parseAlign
import com.github.quadflask.react.navermap.ReactUtil.parseColorString
import com.github.quadflask.react.navermap.ReactUtil.toNaverLatLng
import com.naver.maps.map.overlay.Marker
import kotlin.math.roundToInt

open class RNNaverMapMarkerManager(reactContext: ReactApplicationContext) : EventEmittableViewGroupManager<RNNaverMapMarker?>(reactContext) {
  override val eventNames = arrayOf("onClick")

  override fun getName() = "RNNaverMapMarker"

  override fun createViewInstance(reactContext: ThemedReactContext): RNNaverMapMarker {
    return RNNaverMapMarker(this, reactContext)
  }

  @ReactProp(name = "coordinate")
  fun setCoordinate(view: RNNaverMapMarker, map: ReadableMap?) {
    view.setCoordinate(toNaverLatLng(map))
  }

  @ReactProp(name = "anchor")
  fun setAnchor(view: RNNaverMapMarker, map: ReadableMap?) {
    view.setAnchor(ReactUtil.toPointF(map, Marker.DEFAULT_ANCHOR))
  }

  @ReactProp(name = "image")
  fun setImage(view: RNNaverMapMarker, source: String?) {
    view.setImage(source)
  }

  @ReactProp(name = "pinColor", defaultInt = Color.RED, customType = "Color")
  fun setPinColor(view: RNNaverMapMarker, pinColor: Int) {
    view.setIconTintColor(pinColor)
  }

  @ReactProp(name = "rotation", defaultFloat = 0.0f)
  fun setMarkerRotation(view: RNNaverMapMarker, rotation: Float) {
    view.rotation = rotation
  }

  @ReactProp(name = "flat", defaultBoolean = false)
  fun setFlat(view: RNNaverMapMarker, flat: Boolean) {
    view.setFlat(flat)
  }

  @ReactProp(name = "width", defaultFloat = 64f)
  fun setWidth(view: RNNaverMapMarker, width: Float) {
    val widthInScreenPx: Int = (metrics.density * width).roundToInt()
    view.width = widthInScreenPx
  }

  @ReactProp(name = "height", defaultFloat = 64f)
  fun setHeight(view: RNNaverMapMarker, height: Float) {
    val heightInScreenPx: Int = (metrics.density * height).roundToInt()
    view.height = heightInScreenPx
  }

  @ReactProp(name = "animated", defaultBoolean = false)
  fun setAnimated(view: RNNaverMapMarker, animated: Boolean) {
    view.setAnimated(animated)
  }

  @ReactProp(name = "easing", defaultInt = -1)
  fun setEasing(view: RNNaverMapMarker, easingFunction: Int) {
    view.setEasing(easingFunction)
  }

  @ReactProp(name = "duration", defaultInt = 500)
  fun setDuration(view: RNNaverMapMarker, duration: Int) {
    view.setDuration(duration)
  }

  @ReactProp(name = "alpha", defaultFloat = 1f)
  fun setAlpha(view: RNNaverMapMarker, alpha: Float) {
    view.alpha = alpha
  }

  @ReactProp(name = "zIndex", defaultInt = 0)
  fun setZIndex(view: RNNaverMapMarker, zIndex: Int) {
    view.setZIndex(zIndex)
  }

  @ReactProp(name = "caption")
  fun setCaption(view: RNNaverMapMarker, map: ReadableMap?) {
    if (map == null || !map.hasKey("text")) {
      view.removeCaption()
      return
    }
    val text = map.getString("text") ?: ""
    val textSize = if (map.hasKey("textSize")) map.getInt("textSize") else 16
    val color = if (map.hasKey("color")) parseColorString(map.getString("color")) else Color.BLACK
    val haloColor = if (map.hasKey("haloColor")) parseColorString(map.getString("haloColor")) else Color.WHITE
    val align = if (map.hasKey("align")) parseAlign(map.getInt("align")) else Marker.DEFAULT_CAPTION_ALIGNS[0]
    view.setCaption(text, textSize, color, haloColor, align)
  }

  // react component render listener
  @ReactProp(name = "tick", defaultInt = 0)
  fun setTick(view: RNNaverMapMarker, tick: Int) {
    Log.e("RNN", "onTick $tick")
    view.updateCustomView()
    Handler(Looper.getMainLooper()).postDelayed({ view.updateCustomView() }, 500) // temporary
  }

  override fun addView(parent: RNNaverMapMarker?, child: View?, index: Int) {
    parent?.addView(child!!, index)
  }

  override fun removeView(parent: RNNaverMapMarker?, view: View?) {
    parent?.removeView(view!!)
  }

/*
  override fun createShadowNodeInstance(): LayoutShadowNode {
    // A custom shadow node is needed in order to pass back the width/height of the map to the
    // view manager so that it can start applying camera moves with bounds.
    return SizeReportingShadowNode()
  }

  // TODO test SizeReportingShadowNode
  override fun updateExtraData(view: RNNaverMapMarker, extraData: Any) {
    // This method is called from the shadow node with the width/height of the rendered
    // marker view.
    val data = extraData as Map<String, Float>
    val width = data["width"]!!
    val height = data["height"]!!
    Log.e("RNN", "RNNaverMapMarkerManager.updateExtraData(view: ${view.id}, data: ${width.toInt()}x${height.toInt()})")
    //view.update(width.toInt(), height.toInt())
  }
*/
}
